File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step2/D/D.py
class DSU(object):
    def __init__(self, n: int) -> None:
        self.p = list(range(n + 1))
        self.d = [0] * (n + 1)
        return
    
    def __get(self, a: int) -> tuple[int, int]:
        if self.p[a] == a:
            return (a, 0)
        par, depth = self.__get(self.p[a])
        self.p[a] = par
        self.d[a] += depth
        return (self.p[a], self.d[a])

    def unite(self, a: int, b: int) -> None:
        self.p[a] = b
        self.d[a] += 1
        return
    
    def get_d(self, c: int) -> int:
        return self.__get(c)[1]

def main() -> None:
    n, m = map(int, input().split())

    dsu = DSU(n)
    for _ in range(m):
        a = list(map(int, input().split()))
        if a[0] == 1:
            dsu.unite(a[1], a[2])
        else: # a[0] == 2
            print(dsu.get_d(a[1]))
    return

if __name__ == '__main__':
    main()

File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step2/D/D.cpp
// author: Danila "akshin_" Axyonov

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
typedef long long int lli;

#define plli                pair<lli, lli>
#define mp(_first, _second) make_pair(_first, _second)
#define pb(_elem)           push_back(_elem)




class DSU {
private:
    vector<lli> p;
    vector<lli> d;

    plli __get(lli a) {
        if (this->p[a] == a)
            return mp(a, 0LL);
        auto [par, depth] = this->__get(this->p[a]);
        this->p[a] = par;
        this->d[a] += depth;
        return mp(this->p[a], this->d[a]);
    }

public:
    DSU(lli n) {
        this->p.resize(n + 1LL);
        generate(this->p.begin(), this->p.end(), []() { static lli i = 0LL; return i++; });
        this->d.assign(n + 1LL, 0LL);
        return;
    }

    void unite(lli a, lli b) {
        this->p[a] = b;
        ++this->d[a];
        return;
    }

    lli get_d(lli a) {
        return this->__get(a).second;
    }
};




void solve();

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    solve();
    
    return 0;
}

void solve() {
    lli n,
        m;
    cin >> n >> m;

    DSU dsu(n);
    while (m--) {
        lli cmd;
        cin >> cmd;
        if (cmd == 1LL) {
            lli a,
                b;
            cin >> a >> b;
            dsu.unite(a, b);
        } else { // cmd == 2LL
            lli c;
            cin >> c;
            cout << dsu.get_d(c) << '\n';
        }
    }
    
    return;
}
